File: java/src/test/java/it/giannibombelli/wsc2026/payment/application/usecases/RefundRequestingTest.java
package it.giannibombelli.wsc2026.payment.application.usecases;

import it.giannibombelli.wsc2026.common.domain.identity.EntityId;
import it.giannibombelli.wsc2026.common.domain.primitive.ClientReference;
import it.giannibombelli.wsc2026.common.domain.primitive.Money;
import it.giannibombelli.wsc2026.common.domain.primitive.Timestamp;
import it.giannibombelli.wsc2026.payment.application.commands.RefundTransaction;
import it.giannibombelli.wsc2026.payment.domain.events.PaymentEvent;
import it.giannibombelli.wsc2026.payment.domain.events.RefundRequested;
import it.giannibombelli.wsc2026.payment.domain.payment.Payment;
import it.giannibombelli.wsc2026.payment.domain.payment.PaymentId;
import it.giannibombelli.wsc2026.payment.domain.payment.Provider;
import it.giannibombelli.wsc2026.payment.domain.payment.ProviderReference;
import it.giannibombelli.wsc2026.payment.domain.payment.TransactionId;
import it.giannibombelli.wsc2026.payment.domain.ports.PaymentRepository;
import it.giannibombelli.wsc2026.payment.infrastructure.SqlitePaymentRepository;
import it.giannibombelli.wsc2026.testsupport.DatabaseSetup;
import it.giannibombelli.wsc2026.testsupport.events.CapturingEventPublisher;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;

import javax.sql.DataSource;
import java.math.BigDecimal;
import java.time.Instant;
import java.util.UUID;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

class RefundRequestingTest {
    private PaymentRepository repository;
    private CapturingEventPublisher<PaymentEvent> publisher;
    private RefundRequesting refundRequesting;

    @BeforeEach
    void setUp() {
        final DataSource dataSource = DatabaseSetup.initializeInMemoryDb("payment");
        repository = new SqlitePaymentRepository(dataSource);
        publisher = new CapturingEventPublisher<>();
        refundRequesting = new RefundRequesting(repository, publisher);
    }

    @Nested
    class Construction {
        @Test
        void rejectNullRepository() {
            assertThatThrownBy(() -> new RefundRequesting(null, publisher))
                .isInstanceOf(NullPointerException.class);
        }

        @Test
        void rejectNullPublisher() {
            assertThatThrownBy(() -> new RefundRequesting(repository, null))
                .isInstanceOf(NullPointerException.class);
        }
    }

    @Nested
    class Execution {
        @Test
        void rejectNullCommand() {
            assertThatThrownBy(() -> refundRequesting.invoke(null))
                .isInstanceOf(IllegalArgumentException.class);
        }

        @Test
        void shouldPublishRefundRequested() {
            PaymentId paymentId = seedAcceptedPayment(new BigDecimal("50.00"));
            Money refundAmount = new Money(new BigDecimal("50.00"));

            RefundRequested event = refundRequesting.invoke(new RefundTransaction(paymentId, refundAmount));

            assertThat(event.aggregateId()).isEqualTo(paymentId);
            assertThat(event.amount()).isEqualTo(refundAmount);
            assertThat(publisher.events()).hasSize(1);
            assertThat(publisher.events().get(0)).isInstanceOf(RefundRequested.class);
        }

        @Test
        void shouldFailIfPaymentNotFound() {
            PaymentId paymentId = EntityId.generate(PaymentId::new);
            Money refundAmount = new Money(new BigDecimal("10.00"));

            assertThatThrownBy(() -> refundRequesting.invoke(new RefundTransaction(paymentId, refundAmount)))
                .isInstanceOf(IllegalArgumentException.class);
        }

        @Test
        void shouldFailIfPaymentNotAccepted() {
            PaymentId paymentId = EntityId.generate(PaymentId::new);
            UUID clientReference = UUID.randomUUID();
            Money amount = new Money(new BigDecimal("50.00"));
            Payment payment = Payment.request(paymentId, new ClientReference(clientReference), amount, new Timestamp(Instant.now()));
            repository.save(payment);

            Money refundAmount = new Money(new BigDecimal("10.00"));
            assertThatThrownBy(() -> refundRequesting.invoke(new RefundTransaction(paymentId, refundAmount)))
                .isInstanceOf(IllegalStateException.class);
        }

        private PaymentId seedAcceptedPayment(BigDecimal amount) {
            PaymentId paymentId = EntityId.generate(PaymentId::new);
            UUID clientReference = UUID.randomUUID();
            Money money = new Money(amount);
            Instant requestedAt = Instant.now();
            Payment payment = Payment.request(paymentId, new ClientReference(clientReference), money, new Timestamp(requestedAt));
            TransactionId transactionId = EntityId.generate(TransactionId::new);
            payment.startTransaction(
                transactionId,
                Provider.PAYPAL,
                new ProviderReference(UUID.randomUUID()),
                money,
                new Timestamp(requestedAt.plusSeconds(1))
            );
            payment.acceptTransaction(transactionId, new Timestamp(requestedAt.plusSeconds(60)));
            repository.save(payment);
            return paymentId;
        }
    }
}

File: typescript/tests/payment/application/usecases/refundRequesting.test.ts
import { beforeEach, describe, expect, it } from 'vitest';
import { generateId } from '@/common/domain/identity/entityId.js';
import { ClientReference } from '@/common/domain/primitive/clientReference.js';
import { Money } from '@/common/domain/primitive/money.js';
import { Timestamp } from '@/common/domain/primitive/timestamp.js';
import { RefundTransaction, refundTransaction } from '@/payment/application/commands/refundTransaction.js';
import { RefundRequesting } from '@/payment/application/usecases/refundRequesting.js';
import { PaymentEvent } from '@/payment/domain/events/paymentEvent.js';

import { Payment } from '@/payment/domain/payment/payment.js';
import { PaymentId } from '@/payment/domain/payment/paymentId.js';
import { Provider } from '@/payment/domain/payment/provider.js';
import { ProviderReference } from '@/payment/domain/payment/providerReference.js';
import { TransactionId } from '@/payment/domain/payment/transactionId.js';
import { Uuid } from '@/common/domain/primitive/uuid.js';
import { SqlitePaymentRepository } from '@/payment/infrastructure/sqlitePaymentRepository.js';
import { CapturingEventPublisher } from '../../../testsupport/events/capturingEventPublisher.js';
import { DatabaseSetup } from '../../../testsupport/databaseSetup.js';

describe('RefundRequesting', () => {
  let repository: SqlitePaymentRepository;
  let publisher: CapturingEventPublisher<PaymentEvent>;
  let refundRequesting: RefundRequesting;

  beforeEach(() => {
    const database = DatabaseSetup.initializeInMemoryDb('payment');
    repository = new SqlitePaymentRepository(database);
    publisher = new CapturingEventPublisher();
    refundRequesting = new RefundRequesting(repository, publisher);
  });

  describe('construction', () => {
    it('rejects null repository', () => {
      expect(() => new RefundRequesting(null as unknown as SqlitePaymentRepository, publisher)).toThrow();
    });

    it('rejects null publisher', () => {
      expect(() => new RefundRequesting(repository, null as unknown as CapturingEventPublisher<PaymentEvent>)).toThrow();
    });
  });

  describe('execution', () => {
    it('rejects null command', () => {
      expect(() => refundRequesting.invoke(null as unknown as RefundTransaction)).toThrow();
    });

    it('should publish refund requested', () => {
      const paymentId = seedAcceptedPayment(50);
      const refundAmount = new Money(50);

      const event = refundRequesting.invoke(refundTransaction(paymentId, refundAmount));

      expect(event.aggregateId).toEqual(paymentId);
      expect(event.amount).toEqual(refundAmount);
      expect(publisher.events()).toHaveLength(1);
      expect(publisher.events()[0].kind).toBe('RefundRequested');
    });

    it('should fail if payment not found', () => {
      const paymentId = generateId((value) => new PaymentId(value));
      const refundAmount = new Money(10);

      expect(() => refundRequesting.invoke(refundTransaction(paymentId, refundAmount))).toThrow();
    });

    it('should fail if payment not accepted', () => {
      const paymentId = generateId((value) => new PaymentId(value));
      const payment = Payment.request(
        paymentId,
        new ClientReference(Uuid.fromString(crypto.randomUUID())),
        new Money(50),
        Timestamp.now()
      );
      repository.save(payment);

      const refundAmount = new Money(10);
      expect(() => refundRequesting.invoke(refundTransaction(paymentId, refundAmount))).toThrow();
    });
  });

  function seedAcceptedPayment(amount: number): PaymentId {
    const paymentId = generateId((value) => new PaymentId(value));
    const clientReference = crypto.randomUUID();
    const money = new Money(amount);
    const requestedAt = Timestamp.now();
    const payment = Payment.request(paymentId, new ClientReference(Uuid.fromString(clientReference)), money, requestedAt);
    const transactionId = generateId((value) => new TransactionId(value));
    payment.startTransaction(
      transactionId,
      Provider.PAYPAL,
      new ProviderReference(Uuid.generate()),
      money,
      new Timestamp(new Date(requestedAt.value.getTime() + 1000))
    );
    payment.acceptTransaction(transactionId, new Timestamp(new Date(requestedAt.value.getTime() + 60000)));
    repository.save(payment);
    return paymentId;
  }
});
